File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <bits/stdc++.h>

using namespace std;

int getCount(vector<vector<int>>& wires, vector<bool>& visit, int n, int cutEdge) {
    int cnt = 0;
    deque<int> dq;
    for (int i = 1; i <= n; i++) {
        if (!visit[i]) {
            visit[i] = true;
            dq.push_back(i);
            cnt++;
            break;
        }
    }
    while (!dq.empty()) {
        int curr = dq.front();
        dq.pop_front();
        for (int i = 0; i < wires.size(); i++) {
            if (i == cutEdge) continue;
            if (!visit[wires[i][0]] && visit[wires[i][1]]) {
                visit[wires[i][0]] = true;
                dq.push_back(wires[i][0]);
                cnt++;
            }
            if (!visit[wires[i][1]] && visit[wires[i][0]]) {
                visit[wires[i][1]] = true;
                dq.push_back(wires[i][1]);
                cnt++;
            }
        }
    }
    return cnt;
}

int bfs(int n, vector<vector<int>>& wires, int cutEdge) {
    vector<bool> visit(n+1);
    int cnt1 = getCount(wires, visit, n, cutEdge);
    int cnt2 = getCount(wires, visit, n, cutEdge);
    return abs(cnt1 - cnt2);
}

int solution(int n, vector<vector<int>> wires) {
    int minVal = 101;
    for (int i = 0; i < wires.size(); i++) {
        int cnt = bfs(n, wires, i);
        if (minVal > cnt) minVal = cnt;
    }
    return minVal;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    private int[] parent;
    
    private int find(int x) {
        if (parent[x] == x) return x;
        return find(parent[x]);
    }
    
    private boolean union(int x, int y) {
        x = find(x);
        y = find(y);
        
        if (x == y) return false;
        if (x <= y) parent[y] = x;
        else parent[x] = y;
        return true;
    }
    
    private int count() {
        int x = find(1), cnt1 = 1, cnt2 = 0;
        for (int i = 2; i < parent.length; i++) {
            if (x == find(i)) cnt1++;
            else cnt2++;
        }
        return Math.abs(cnt1 - cnt2);
    }
    
    public int solution(int n, int[][] wires) {
        int diff = n;
        for (int i = 0; i < wires.length; i++) {
            parent = new int[n+1];
            for (int j = 1; j <= n; j++)
                parent[j] = j;
            for (int j = 0; j < wires.length; j++) {
                if (i == j) continue;
                union(wires[j][0], wires[j][1]);
            }
            int res = count();
            if (res < diff) diff = res;
        }
        return diff;
    }
}